tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -29,84 +39,189 @@
+ *    There may be a case when a server pulls all load from all other servers.
+ *    And such condition (if it occurs although it is quite improbable) is
+ *    quite stable: it cannot be fixed by adding/removing servers and restarting
+ *    the Tempesta FW instance.
+ *  - For every HTTP request, we have to scan the list of all servers to find
+ *    a matching one with the highest weight. That adds some overhead.
+ *    Currently the linear search is used. We may switch to binary search, but
+ *    benchmarks are required to prove that it will behave better on the small
+ *    number of servers that we usually have.
+ */
 static TfwServer *
 tfw_sched_hash_get_srv(TfwMsg *msg)
 {
-	TfwServer *srv;
-	int n;
+	TfwServer *best_srv;
+	TfwSrvHash *curr_srv_hash;
+	TfwSrvHashList *srv_hash_list;
+	unsigned long msg_hash, curr_weight, best_weight;
+
+	msg_hash = tfw_http_req_key_calc((TfwHttpReq *)msg);
 
-	unsigned long hash = tfw_http_req_key_calc((TfwHttpReq *)msg);
+	rcu_read_lock();
+	srv_hash_list = rcu_dereference(tfw_srv_hash_list);
 
-	do {
-		n = servers_n;
-		if (!n) {
-			ERR("No servers added to the scheduler\n");
-			return NULL;
+	/* 1. Set best = first element of the list. */
+	curr_srv_hash = &srv_hash_list->srv_hashes[0];
+	best_srv = curr_srv_hash->srv;
+	best_weight = msg_hash ^ curr_srv_hash->hash;
+
+	/* 2. Try to find a better one among 2nd...Nth elements of the list.
+	 * We don't check for N == 0 here; the list is terminated with two NULL
+	 * elements when it is empty, so we can avoid the extra branch here. */
+	while ((++curr_srv_hash)->srv) {
+		curr_weight = msg_hash ^ curr_srv_hash->hash;
+		if (curr_weight > best_weight) {
+			best_weight = curr_weight;
+			best_srv = curr_srv_hash->srv;
 		}
-		srv = servers[hash % n];
-	} while (!srv);
+		++curr_srv_hash;
+	}
+
+	rcu_read_unlock();
 
-	return srv;
+	return best_srv;
+}
+
+static unsigned long
+tfw_sched_hash_calc_srv(TfwServer *srv)
+{
+	unsigned long hash;
+	union {
+		TfwAddr addr;
+		unsigned char bytes[0];
+	} a;
+	size_t i, bytes_n;
+
+	/**
+	 * Here we just cast the whole TfwAddr to an array of bytes
+	 * and use the standard hash_long() to calculate the hash.
+	 *
+	 * That only works if the following invariants are held:
+	 *  - There are no gaps between structure fields.
+	 *  - No structure fields (e.g. sin6_flowinfo) are changed if we
+	 *    re-connect to the same server.
+	 */
+	tfw_server_get_addr(srv, &a.addr);
+	bytes_n = tfw_addr_sa_len(&a.addr);
+
+	hash = 0;
+	for (i = 0; i < bytes_n; ++i) {
+		hash = hash_long(hash ^ a.bytes[i], BITS_PER_LONG);
+	}
+
+	return hash;
 }
 
 static int
 tfw_sched_hash_add_srv(TfwServer *srv)
 {
 	int ret = 0;
+	TfwSrvHash *new_list_entry;
+	TfwSrvHashList *new_list, *old_list;
+	size_t new_list_size, old_list_size;
+
+	spin_lock_bh(&srv_hashes_update_lock);
+
+	old_list = tfw_srv_hash_list;
+	old_list_size = TFW_SRV_HASH_LIST_SIZE(old_list->n);
+	new_list_size = TFW_SRV_HASH_LIST_SIZE(old_list->n + 1);
```

Comment:
Usually growing in exponentionaly is more preferred than realloc and copy the array for each server. Just keep in ming #76.